### DISCUSSION TITLE
Is it possible to print the AST produced by typescript-estree?

### DISCUSSION THREAD

**Discussion by tmcw** [typescript-eslint/typescript-eslint#6377] [Technical Discussions]
Hi! I'm working on a project that modifies TypeScript code and tries to print the result. The `typescript-estree` project which is a part of `typescript-eslint` seems extremely useful in this regard, because it'd let me deal with the AST with tools that interact with ESTree, which is very cool!

Where I'm having some issues is how to get the updated AST out. Prettier uses typescript-estree internally, but it does some internal modification of the AST before trying to print it. Tools that print ESTree structures like `astring` seem to ignore the types that `typescript-estree` tries to maintain.

Has anyone run into this issue and found a way through? It's turning out to be pretty hard to do transformations of TypeScript.

Thanks!

**Comment by bradzacher** [comment]
`astring` only supports ESTree. But our parser produces an extended form of ESTree (we call it TS-ESTree).

If you want to print it, you need a tool capable of printing TS-ESTree - I only know of one myself - prettier.
Even then it's a real pain to work with prettier here because it's not made to work with custom parsers too well.

It's something that's doable though! While working at meta I worked on a tool to do ast-based codemods of flow code. We used prettier to print the Flow-ESTree AST out.

This is the code used to do that, which could be modified to work with TS-ESTree
https://github.com/facebook/hermes/blob/main/tools/hermes-parser/js/hermes-transform/src/transform/print.js

**Reply by tmcw** [reply]
Thanks! Was afraid I was missing something simple. It's been surprising to see how tricky it is to parse, modify, and stringify TypeScript code. But typescript-estree is a great tool in that arsenal, and hopefully it'll be possible to do something like the hermes project.

**Reply by pgsandstrom** [reply]
Did you have any luck with converting TSESTree to code? Im in the same situation here and having some troubles getting anything to work.
